src/batHashtbl.mlv

```diff
@@ -357,10 +362,108 @@ let filter_map_inplace f h =
        | Some v' -> Cons (k, v', loop next)) in
   BatArray.modify loop hc.data
 
-(*$= filter_map_inplace & ~printer:(IO.to_string (List.print Int.print))
+(*$= filter_map_inplace & ~printer
   (let h = Enum.combine (1 -- 5, 1 -- 5) |> of_enum in \
    filter_map_inplace (fun _ x -> if x>3 then Some (x+1) else None) h ; \
-   values h |> List.of_enum |> List.sort Int.compare) [5; 6]
+   to_comparable h) [5; 6]
+*)
+
+
+let merge f a b =
+  let res = create (max (length a) (length b)) in
+  let may_add_res k v1 v2 =
+    BatOption.may (add res k) (f k v1 v2) in
+  iter (fun k v1 ->
+    match find b k with
+    | exception Not_found ->
+        may_add_res k (Some v1) None
+    | v2 ->
+        may_add_res k (Some v1) (Some v2)
+  ) a ;
+  iter (fun k v2 ->
+    match find a k with
+    | exception Not_found ->
+        may_add_res k None (Some v2)
+    | _ -> () (* done above *)
+  ) b ;
+  res
+
+(*$= merge & ~printer
+  [] \
+    (let a = create 0 and b = create 0 in \
+     merge (fun k _ _ -> Some k) a b |> to_comparable)
+  [1; 2; 3; 4; 5] \
+    (let a = Enum.combine (1 -- 5, 1 -- 5) |> of_enum and b = create 0 in \
+     merge (fun _ l _ -> l) a b |> to_comparable)
+  [] \
+    (let a = Enum.combine (1 -- 5, 1 -- 5) |> of_enum and b = create 0 in \
+     merge (fun _ _ r -> r) a b |> to_comparable)
+  [] \
+    (let a = create 0 and b = Enum.combine (1 -- 5, 1 -- 5) |> of_enum in \
+     merge (fun _ l _ -> l) a b |> to_comparable)
+  [1; 2; 3; 4; 5] \
+    (let a = create 0 and b = Enum.combine (1 -- 5, 1 -- 5) |> of_enum in \
+     merge (fun _ _ r -> r) a b |> to_comparable)
+  [1; 2; 3] \
+    (let a = Enum.combine (1 -- 3, 1 -- 3) |> of_enum and b = Enum.combine (3 -- 6, 13 -- 15) |> of_enum in \
+     merge (fun _ l _ -> l) a b |> to_comparable)
+  [13; 14; 15] \
+    (let a = Enum.combine (1 -- 3, 1 -- 3) |> of_enum and b = Enum.combine (3 -- 5, 13 -- 15) |> of_enum in \
+     merge (fun _ _ r -> r) a b |> to_comparable)
+  [] \
+    (let a = Enum.combine (1 -- 3, 1 -- 3) |> of_enum and b = Enum.combine (3 -- 5, 3 -- 5) |> of_enum in \
+     merge (fun _ _ _ -> None) a b |> to_comparable)
```

Comment:
`merge` can express both `inter` and `union`, and it would be nice to see those appear in tests. We could consider having qcheck-style property tests (`equal (inter a a) a`, `equal (union a a) a`, `equal (union a empty) a`, `equal (inter a empty) empty`, `equal (inter a b) (inter b a)`, `equal (union a b) (union b a)`, etc.

Reply:
I gave it a try but I'm not very satisfied with it. Please have a look.
The problem is that it gives a misleading feeling of confidence, but actually the union and intersection of hash tables when multiple bindings to the same keys is allowed are not as natural as it seems. Indeed, what's the intersection of `{1->2}` and `{1->3}` ? Is it `{1->2}`, `{1->3}`, `{1->[2;3]}` or `{}` ? And what's the union of `{1->1}` with `{1->2}` ? Is it `{1->[1; 2]}` ? but then, what's the union of `{1->1}` and `{1->1}` ? is it `{1->[1;1]}` ?

Implementing those tests actually made me realise that `Hashtbl.iter` and `Hashtbl.keys` were not working as I've always assumed they would. For instance, I've always though `iter` would send multiple bindings to the same key successively, but it does not (it iterate over the bucket but keys might not be next to each others in the bucket), and I've also always though keys would return the keys only once but it actually return the keys several times in case of multiple bindings. So, quite useful  :)

At the end of the day (literally), I updated the documentation of [merge] to guard against using it in case of multiple bindings.

WDYT?